### PR TITLE
chore: Upgrade to hugr 0.8.0

### LINKED ISSUE #63: Upgrade hugr to 0.8.0
Hugr 0.8.0 is a non-trivial upgrade because it introduces the idea that all ops should carry their extensions in their signatures. This issue exists as a stepping stone to get us targeting the latest version of hugr

### PR DESCRIPTION
**Pull request by croyzor**
Resolves #63

### CODE REVIEW

File: brat/Brat/Compile/Hugr.hs [hunk 1]
```diff
@@ -201,7 +200,7 @@ compileConst parent tm ty = do
   constId <- addNode "Const" (OpConst (ConstOp parent (valFromSimple tm)))
   loadId <- case ty of
     HTFunc poly@(PolyFuncType [] _) ->
-      addNode "LoadFunction" (OpLoadFunction (LoadFunctionOp parent poly [] (FunctionType [] [HTFunc poly])))
+      addNode "LoadFunction" (OpLoadFunction (LoadFunctionOp parent poly [] (FunctionType [] [HTFunc poly] [])))
```
Comment (mark-koch) [inline brat/Brat/Compile/Hugr.hs:203]: Why are the extension reqs for `LoadFunction` empty?

File: brat/Brat/Compile/Hugr.hs [hunk 2]
```diff
@@ -295,13 +294,14 @@ compileClauses parent ins ((matchData, rhs) :| clauses) = do
   didntMatch outTys parent ins = case nonEmpty clauses of
     Just clauses -> compileClauses parent ins clauses
     -- If there are no more clauses left to test, then the Hugr panics
-    Nothing -> let sig = FunctionType (snd <$> ins) outTys in
-      addNodeWithInputs "Panic" (OpCustom (CustomOp parent "brat" "panic" sig [])) ins outTys
+    -- TODO: This should just be [BRAT] - would that work?
```
Comment (mark-koch) [inline brat/Brat/Compile/Hugr.hs:297]: Don't ops add their own extension as a requirement implicitly?

File: brat/Brat/Compile/Hugr.hs [hunk 3]
```diff
@@ -437,7 +438,7 @@ compileWithInputs parent name = gets compiled >>= (\case
           _ -> compileWithInputs parent outNode >>= \case
             Just calleeId -> do
               callerId <- addNode ("indirect_call(" ++ show calleeId ++ ")")
-                                  (OpCallIndirect (CallIndirectOp parent (FunctionType ins outs)))
+                                  (OpCallIndirect (CallIndirectOp parent (FunctionType ins outs bratExts {-[]-})))
```
Comment (mark-koch) [inline brat/Brat/Compile/Hugr.hs:441]: Should this be empty or not, c.f. `LoadFunction`?

File: brat/Brat/Compile/Hugr.hs [hunk 4]
```diff
@@ -820,6 +823,8 @@ compileModule venv = do
     put st { compiled = M.empty }
     body)
  where
+  -- HELP! This doesn't work at all when extensions are involved!!!
+  --
```
Comment (mark-koch) [inline brat/Brat/Compile/Hugr.hs:827]: Still looking for help? What's going wrong?

File: brat/Data/Hugr.hs [hunk 5]
```diff
@@ -13,6 +13,19 @@ import Data.Text (Text, pack)
...
+-- but instead we'll overapproximate.
+bratExts :: [ExtensionId]
+bratExts =
+ ["prelude"
+ ,"arithmetic.int_ops"
+ ,"arithmetic.float_ops"
+ ,"collections"
+ ,"logic"
+ ,"tket2.quantum"
+ ,"BRAT"]
```
Comment (mark-koch) [inline brat/Data/Hugr.hs:26]: `arithmetic.int_types` and `arithmetic.float_types` are missing

File: brat/Data/Hugr.hs [hunk 6]
```diff
@@ -472,15 +487,13 @@ substOp :: node
...
         -> CustomOp node
 substOp parent outerSig innerSigs
-  = CustomOp parent "Brat" "Substitute" sig args
+  = CustomOp parent "BRAT" "Substitute" sig [toArg outerSig, TASequence (toArg <$> innerSigs)]
  where
-  sig = FunctionType (toFunc <$> (outerSig : innerSigs)) [toFunc outerSig]
-  args = [funcToSeq outerSig, TASequence (funcToSeq <$> innerSigs)]
-
-  funcToSeq (FunctionType ins outs) = TASequence [toSeq ins, toSeq outs]
+  sig = FunctionType (toFunc <$> (outerSig : innerSigs)) [toFunc outerSig] ["BRAT"]
```
Comment (mark-koch) [inline brat/Data/Hugr.hs:492]: Do we also need to include the reqs of the input graphs?

File: brat/Brat/Compile/Hugr.hs (hunk 1 above)
Comment (croyzor) [reply brat/Brat/Compile/Hugr.hs:203]: LoadFunction is only used for const functions i thought? The extensions in the function that's loaded are boxed up until it is `call`ed

File: brat/Brat/Compile/Hugr.hs (hunk 3 above)
Comment (croyzor) [reply brat/Brat/Compile/Hugr.hs:441]: The extensions here should match the extensions of the callee, but we're overapproximating

File: brat/Brat/Compile/Hugr.hs (hunk 1 above)
Comment (mark-koch) [reply brat/Brat/Compile/Hugr.hs:203]: I have no idea, just checking 😅

But that makes sense 👍

**Review by mark-koch — APPROVED** [review]

**Pull request merged**